### PR TITLE
Sam monitoring

### PR DESCRIPTION
**Pull request by sjkoelle**
Contains minor reworking of diagonalness computation (move from utils to monitoring and accepting output length argument).

### CODE REVIEW

File: applications/inference_monitoring_streamlit.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
+import streamlit as st
+import torch
+from uberduck_ml_dev.monitoring.generate import _get_inference
+from uberduck_ml_dev.models.tacotron2 import DEFAULTS as TACOTRON2_DEFAULTS
+from uberduck_ml_dev.models.tacotron2 import Tacotron2
```
Comment (williamluer) [inline applications/inference_monitoring_streamlit.py:5]: Could move to one line:
```python
from uberduck_ml_dev.models.tacotron2 import Tacotron2, DEFAULTS as TACOTRON2_DEFAULTS
```

File: applications/inference_monitoring_streamlit.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            hparams.gate_threshold = gate_threshold
+            if n_speakers > 1:
+                hparams.has_speaker_embedding = True
+            model = Tacotron2(hparams)
+            device = 'cuda'
+            model = Tacotron2(hparams)
+            if model_format == 'OD':
+                model.from_pretrained(model_dict=torch.load(model_path), device=device)
+            if model_format == 'D':
+                model.from_pretrained(warm_start_path=model_path, device=device)
```
Comment (williamluer) [inline applications/inference_monitoring_streamlit.py:213]: Might want to introduce cuda as a parameter

File: applications/inference_monitoring_streamlit.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    
+    model_path = st.text_input("Add model path","/mnt/disks/uberduck-experiments-v0/gcs_logbucket/tensorboard-ubstudio_v3_20epochcheckpoints/model/tacotron2_20.pt")
+    if st.button("Add model"):
+        get_model().append(model_path)
+    model_path = st.selectbox('Model', unique(get_model()))
+
+    vc_path = st.text_input("Add vocoder config path","/mnt/disks/uberduck-experiments-v0/uberduck-ml-dev/nbs/test/fixtures/models/config_v1.json")
+    if st.button("Add vocoder config"):
+        get_vc().append(vc_path)
+    vc_path = st.sidebar.selectbox('Vocoder config', unique(get_vc()))
```
Comment (williamluer) [inline applications/inference_monitoring_streamlit.py:161]: These files don't actually exist on uberduck-ml-dev so I'd suggest putting a placeholder like `/path/to/config_v1.json` or `/path/to/tacotron.pt`

File: uberduck_ml_dev/monitoring/generate.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+def _get_inference(model, vocoder, texts, speaker_ids, symbol_set, arpabet):
+
+    text_padded, input_lengths = prepare_input_sequence(
+        texts, cpu_run=False, arpabet=arpabet, symbol_set=symbol_set
+    )
+    input_ = text_padded, input_lengths, speaker_ids
+    output = model.inference(input_)
+    audio = vocoder.infer(output[1][:1])
+    return audio
```
Comment (williamluer) [inline uberduck_ml_dev/monitoring/generate.py:19]: Can you use this function here? https://github.com/uberduck-ai/uberduck-ml-dev/commit/5e87433b0bda51dc6b0193f6f3422d90e23acff7#diff-f8baa94114a0a737cd41dc5522a4397c66a6dfad3a6ed7c9bce1342c00f8ee66R61-R75

File: uberduck_ml_dev/monitoring/statistics.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import torch
+
+
+def get_alignment_metrics(
+    alignments, average_across_batch=True, input_lengths=None, output_lengths=None
+):
+    """See https://github.com/NVIDIA/tacotron2/pull/284,
+    https://github.com/CookiePPP/cookietts/blob/c871f5f7b5790656d5b57bcd9e63946a2da52f0f/CookieTTS/utils/model/utils.py#L59"""
+    alignments = alignments.transpose(1, 2)  # [B, dec, enc] -> [B, enc, dec]
+
```
Comment (williamluer) [inline uberduck_ml_dev/monitoring/statistics.py:15]: Are the correct licenses added to uberduck-ml-dev from these?

File: uberduck_ml_dev/monitoring/streamlit.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    speakers = [st.text_input("Speaker_id", 0)]
+
+    hparams = TACOTRON2_DEFAULTS
+    hparams.n_speakers = n_speakers
+    hparams.gate_threshold = gate_threshold
+    if n_speakers > 1:
+        hparams.has_speaker_embedding = True
+    model = Tacotron2(hparams)
+    device = "cuda"
+    model = Tacotron2(hparams)
```
Comment (williamluer) [inline uberduck_ml_dev/monitoring/streamlit.py:43]: Duplicate lines?

**Review by zachwe — APPROVED** [review]

**Pull request merged**